vector/read.go

```diff
@@ -14,570 +16,597 @@ import (
+		values := make([]uint32, 0, count)
+		for _, segment := range segmap {
+			err := readSegment(readerAt, buf, segment)
 			if err != nil {
-				if err == io.EOF {
-					break
-				} else {
-					return nil, err
-				}
+				return nil, err
+			}
+			it := zcode.Iter(*buf)
+			for !it.Done() {
+				bs := it.Next()
+				values = append(values, uint32(zed.DecodeUint(bs)))
 			}
-			values = append(values, uint32(zed.DecodeUint(bs)))
 		}
-		vector := &uint32s{
+		return &uint32s{
 			values: values,
-		}
-		return vector, nil
+		}, nil
 
 	case zed.TypeUint64:
-		var values []uint64
-		for {
-			bs, err := readBytes()
+		values := make([]uint64, 0, count)
+		for _, segment := range segmap {
+			err := readSegment(readerAt, buf, segment)
 			if err != nil {
-				if err == io.EOF {
-					break
-				} else {
-					return nil, err
-				}
+				return nil, err
+			}
+			it := zcode.Iter(*buf)
+			for !it.Done() {
+				bs := it.Next()
+				values = append(values, uint64(zed.DecodeUint(bs)))
 			}
-			values = append(values, uint64(zed.DecodeUint(bs)))
 		}
-		vector := &uint64s{
+		return &uint64s{
 			values: values,
-		}
-		return vector, nil
+		}, nil
 
 	case zed.TypeNull:
 		return &constants{}, nil
 
 	case zed.TypeType:
-		var values []zed.Type
-		for {
-			bs, err := readBytes()
+		values := make([]zed.Type, 0, count)
+		for _, segment := range segmap {
+			err := readSegment(readerAt, buf, segment)
 			if err != nil {
-				if err == io.EOF {
-					break
-				} else {
-					return nil, err
-				}
+				return nil, err
+			}
+			it := zcode.Iter(*buf)
+			for !it.Done() {
+				bs := it.Next()
+				typ, _ := zctx.DecodeTypeValue(bs)
+				values = append(values, typ)
 			}
-			typ, _ := context.DecodeTypeValue(bs)
-			values = append(values, typ)
 		}
-		vector := &types{
+		return &types{
 			values: values,
-		}
-		return vector, nil
+		}, nil
 
 	default:
 		return nil, fmt.Errorf("unknown VNG type: %T", typ)
 	}
 }
 
-func readInt64s(reader *vngvector.Int64Reader) ([]int64, error) {
-	var ints []int64
-	for {
-		int, err := reader.Read()
-		if err != nil {
-			if err == io.EOF {
-				break
-			} else {
-				return nil, err
-			}
+func readSegment(readerAt io.ReaderAt, buf *[]byte, segment vngvector.Segment) error {
+	*buf = slices.Grow((*buf)[:0], int(segment.MemLength))[:segment.MemLength]
+	return segment.Read(readerAt, *buf)
+}
+
+func readSegmap(readerAt io.ReaderAt, segmap []vngvector.Segment) ([]byte, error) {
+	var memLength int
+	for _, segment := range segmap {
+		memLength += int(segment.MemLength)
+	}
+	data := make([]byte, memLength)
+	offset := 0
+	for _, segment := range segmap {
+		if err := segment.Read(readerAt, data[offset:offset+int(segment.MemLength)]); err != nil {
+			return nil, err
 		}
-		ints = append(ints, int)
+		offset += int(segment.MemLength)
 	}
-	return ints, nil
+	return data, nil
+}
+
+func stripContainers(data []byte, countHint int) ([]byte, []int, error) {
+	offsetFrom := 0
+	offsetTo := 0
```

Comment:
Nit: Use var for zero values.
```suggestion
	var offsetFrom, offsetTo int
```